core/src/main/scala/org/apache/spark/deploy/history/ApplicationCache.scala

```diff
@@ -48,11 +49,28 @@ private[history] class ApplicationCache(
     val retainedApplications: Int,
     val clock: Clock) extends Logging {
 
+  /**
+   * Keep track of SparkUIs in [[ApplicationCache#appCache]] and SparkUIs removed from
+   * [[ApplicationCache#appCache]] but not detached yet.
+   */
+  private val loadedApps = new ConcurrentHashMap[CacheKey, CountDownLatch]()
+
   private val appLoader = new CacheLoader[CacheKey, CacheEntry] {
 
     /** the cache key doesn't match a cached entry, or the entry is out-of-date, so load it. */
     override def load(key: CacheKey): CacheEntry = {
-      loadApplicationEntry(key.appId, key.attemptId)
+      // Ensure old SparkUI has been detached before loading new one.
+      val removalLatch = loadedApps.get(key)
+      if (removalLatch != null) {
+        // Old SparkUI is in the middle of detaching.
+        // Waiting 10 seconds should be enough since detaching usually takes less than 1 second.
+        if (!removalLatch.await(10, TimeUnit.SECONDS)) {
+          throw new TimeoutException("Timed out waiting for old SparkUI to be detached")
```

Comment:
More places in Spark directly use `java.util.concurrent.TimeoutException` instead of `scala.concurrent.TimeoutException`(`s.c.TimeoutException` is just an alias for j.u.c.TimeoutException).

Reply:
Updated